Step-by-step-Tutorials/Failure-Handling.md

```diff
@@ -0,0 +1,59 @@
+
+After you see a grain operation failed you can do one or some  of the following.
+
+- Simply retry your action specially if it doesn't involve any state changes which might be half done.
+- Try to fix the partially changed state which is not stored in storage by calling `DeactivateOnIdle` to kill the activation. It will load the correct state from storage at next call.
+- Kill all related activations as well to ensure a clean state for all of them.
+- Due multi-grain state manipulations using a [Process Manager](https://msdn.microsoft.com/en-us/library/jj591569.aspx) or database transaction to make sure it's either done completely or nothing is changed to avoid the state being messed up,.
+
+Depending on your application the retry logic might follow a simple or complex pattern and you might have to do other stuff like notifying external systems and other things but generally you either have to retry your action, restart the grain/grains involved or stop responding until something which is not available becomes available. 
+
+If you have a grain responsible for database saves and database is not available, you simply have to fail any request until the DB comes back online. If your operation can be simply be retried at user's will like failure of saving a comment in a comment grain, you can just retry when user presses the retry button (until a certain number of times in order to not saturate the network). Specific details of how to do it are really application specific but the possible strategies are the same.
+
+## Strategy parameters and choosing a good strategy
+
+As described in the section above, choosing a strategy is application and context dependent and strategies usually have parameters which again have to be decided at the application level. For example you might want to retry a request maximum 5 times per minute and can deal with it being done eventually but for some other action you might not be able to continue if something is resolved. If your Login grain fails , you can not process any other requests from the user as an example. 
+
+## A note about supervision trees
+
+Developers used to Erlang/Akka/Akka.Net might be surprised to see that there is no supervision tree in Orleans and wonder how they can recover from failures. The point is that in Orleans since actors are reactivated automatically and you don't handle their life-cycle, you usually only deal with killing actors (deactivating them) and not their creation. If you need to store relations between grains, you can simply do it and it is a widely done practice so if my `GameManager`grain fails, all games belong to it should fail as well and the manager can do this for them by calling `Deactivate OnIdle` 
```

Comment:
you currently can't kill a grain externally... Each grain can decide to deactivate itself by calling `this.DeactivateOnIdle()`